envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/filters/RBACFilterFactory.kt

```diff
@@ -0,0 +1,102 @@
+package pl.allegro.tech.servicemesh.envoycontrol.snapshot.listeners.filters
+
+import com.google.protobuf.Any
+import io.envoyproxy.envoy.api.v2.route.HeaderMatcher
+import io.envoyproxy.envoy.config.filter.http.rbac.v2.RBAC as RBACFilter
+import io.envoyproxy.envoy.config.rbac.v2.Permission
+import io.envoyproxy.envoy.config.rbac.v2.Policy
+import io.envoyproxy.envoy.config.rbac.v2.Principal
+import io.envoyproxy.envoy.config.rbac.v2.RBAC
+import pl.allegro.tech.servicemesh.envoycontrol.groups.Group
+import io.envoyproxy.envoy.config.filter.network.http_connection_manager.v2.HttpFilter
+
+import pl.allegro.tech.servicemesh.envoycontrol.groups.Incoming
+import pl.allegro.tech.servicemesh.envoycontrol.groups.IncomingEndpoint
+import pl.allegro.tech.servicemesh.envoycontrol.groups.PathMatchingType
+import pl.allegro.tech.servicemesh.envoycontrol.logger
+import pl.allegro.tech.servicemesh.envoycontrol.snapshot.IncomingPermissionsProperties
+
+class RBACFilterFactory(
+    private val properties: IncomingPermissionsProperties
+) {
+    companion object {
+        private val logger by logger()
+    }
+
+    fun getRules(serviceName: String, incomingPermissions: Incoming): RBAC {
```

Comment:
With roles it changed a bit, so I do not think it will work like this now. I think that mutating state in one small function that does not leak anywhere else is ok.